renormalizer/transport/spectral_function.py

```diff
@@ -0,0 +1,133 @@
+        config when carrying out real time propagation.
+    dump_dir : str
+        the directory for logging and numerical result output.
+    job_name : str
+        the name of the calculation job which determines the file name of the logging and numerical result output.
+    """
+
+    def __init__(
+            self,
+            model: TI1DModel,
+            compress_config: CompressConfig = None,
+            evolve_config: EvolveConfig = None,
+            dump_dir: str = None,
+            job_name: str=None,
+    ):
+        self.model: TI1DModel = model
+        self.compress_config = compress_config
+        if self.compress_config is None:
+            self.compress_config = CompressConfig()
+        # electron-addition Green's function at different $t$ assuming translational invariance
+        self._G_array = []
+        self.e_occupations_array = []
+        self.temperature = Quantity(0)
+        super().__init__(evolve_config, False, dump_dir, job_name)
+
+    @property
+    def G_array(self):
+        """
+        :math:`G_{ij}(t)` as a two dimensional array.
+        The first dimension is :math:`t` and the second dimension is :math:`|i-j|`.
+
+        Returns
+        =======
+        G_array : np.ndarray
+            The Green's function array
+        """
+        return np.array(self._G_array)
+
+    def init_mps(self):
+        creation_oper = Mpo.onsite(self.model, r"a^\dagger", dof_set={self.model.e_dofs[0]})
+        gs = Mps.ground_state(self.model, False)
+        self.h_mpo = Mpo(self.model, offset=Quantity(gs.expectation(Mpo(self.model))))
+        a_ket = creation_oper.apply(gs, canonicalise=True)
+        a_ket.compress_config = self.compress_config
+        a_ket.evolve_config = self.evolve_config
+        a_ket.canonical_normalize()
+        if self.evolve_config.is_tdvp:
+            a_ket = a_ket.expand_bond_dimension(self.h_mpo)
+        return (gs, a_ket)
+
+    def process_mps(self, mps):
+        key = "a"
+        if key not in self.model.mpos:
+            a_opers = [Mpo.onsite(self.model, "a", dof_set={dof}) for dof in self.model.e_dofs]
+            self.model.mpos[key] = a_opers
+        else:
+            a_opers = self.model.mpos[key]
+
+        a_bra_mpo, a_ket_mpo = mps
+        G = a_ket_mpo.expectations(a_opers, a_bra_mpo.conj()) / 1j
+        self._G_array.append(G)
+        self.e_occupations_array.append(a_ket_mpo.e_occupations)
+
+    def evolve_single_step(self, evolve_dt):
+        prev_bra_mpdm, prev_ket_mpdm = self.latest_mps
+        latest_ket_mpdm = prev_ket_mpdm.evolve(self.h_mpo, evolve_dt)
+        return (prev_bra_mpdm, latest_ket_mpdm)
+
+    def get_dump_dict(self):
+        dump_dict = dict()
+        dump_dict['temperature'] = self.temperature.as_au()
+        dump_dict['time series'] = self.evolve_times
+        dump_dict["G array"] = self.G_array
+        ne = self.model.n_edofs
+        ka = np.linspace(0, np.pi, ne // 2 + 1).reshape(1, 1, -1)
```

Comment:
I think ka = np.linspace(-pi,pi,ne+1).reshape(1, 1, -1).
when ne is 2n-1, the result is same as the current definition, when ne is 2n it is different.

Reply:
Actually, it is when ne=2n that the two sets of code are equivalent. I think a better way is to assert ne is even since there's no point looking for the "absent k=0" trouble. Of course, the reason of the striction and the specific choice of the k points should be documented.

Reply:
Yes. I'm wrong. When ne=2n, the results are the same. 
Yes. It is better to assert "ne is even" and add some docs. 
I don't understand the "absent k=0" trouble.

Reply:
In some cases it is desired to analyze spectral function at the k=0 point.

Reply:
The k=0 point is always there. but the k=pi point is not guaranteed (yes when ne=2n, no when ne=2n+1).

Reply:
Yes, you're right. I get confused. Still, I think it's better to make ne an even number, since the k=pi point is also quite special.

Besides, it makes me wondering if the odd/even problem is important in transport problems. I always use odd numbers as a convention from the diffusion dynamics ages.

Reply:
If ne is approaching the thermodynamic limit, I think it is not important for ne to be odd or even.

Reply:
In the end, I decided to make it compatible with even and odd `ne` since in the test case its `ne=3` and `ne=4` is too time consuming.